### PR TITLE
fix: enqueueLinks with redirect urls

### LINKED ISSUE #64: Fix `enqueueLinks` with explicit pseudourls and redirecting urls
Currently the enqueing strategy is always used, we should ignore that if some explicit pseudourls/regexps/globs are provided in the options. 

We also need to rework how the defaultBaseUrl works, we need to have both `request.url` and `request.loadedUrl` in the options, `request.url` being used for enqueing (it is before redirects, which we want for the enqueing base), and `request.loadedUrl` used for the relative links in cheerio crawler.

Lastly, we should not have default for the request queueu in the arguments, it should be always provided either by user or from the context, no default and we should validate that it is there via `ow` in the lowest level implementation.

### PR DESCRIPTION
**Pull request by vladfrangu**
Closes #64

### CODE REVIEW

File: packages/basic-crawler/src/internals/basic-crawler.ts [hunk 1]
```diff
@@ -906,12 +909,8 @@ export class BasicCrawler<
 }
 
 /** @internal */
-export async function basicCrawlerEnqueueLinks(options: BasicCrawlerEnqueueLinksOptions, requestQueue?: RequestQueue) {
-    return enqueueLinks({
-        // FIXME this does not make much sense, as the instance would be ignored by any crawler - the argument needs to be required
-        requestQueue: requestQueue ?? await RequestQueue.open(),
-        ...options,
-    });
+export async function basicCrawlerEnqueueLinks(options: EnqueueLinksOptions) {
```
Comment (B4nan) [inline packages/basic-crawler/src/internals/basic-crawler.ts:912]: let's remove this and use `enqueueLinks` directly

File: test/e2e/test-cheerio-enqueue-links/test.mjs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { Actor } from 'apify';
+import { CheerioCrawler } from '@crawlee/cheerio';
+import deepEqual from 'deep-equal';
+import { expect } from '../tools.mjs';
+
+const crawler = new CheerioCrawler({
+    async requestHandler({ $, enqueueLinks, request, log }) {
+        const { url, loadedUrl } = request;
+
+		const pageTitle = $('title').first().text();
```
Comment (B4nan) [inline test/e2e/test-cheerio-enqueue-links/test.mjs:10]: mixed indent, we should add editorconfig file... :]

File: test/e2e/test-cheerio-enqueue-links/test.mjs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { Actor } from 'apify';
+import { CheerioCrawler } from '@crawlee/cheerio';
+import deepEqual from 'deep-equal';
+import { expect } from '../tools.mjs';
+
+const crawler = new CheerioCrawler({
```
Comment (B4nan) [inline test/e2e/test-cheerio-enqueue-links/test.mjs:6]: lets test the same with one of the browser crawlers too

File: packages/cheerio-crawler/src/internals/cheerio-crawler.ts [hunk 4]
```diff
@@ -1062,23 +1065,39 @@ export class CheerioCrawler<JSONData = unknown> extends BasicCrawler<
...
+    if (options?.baseUrl) {
+        // User provided base url takes priority
+        baseUrl = options.baseUrl;
+    } else if (options?.strategy === EnqueueStrategy.All) {
+        // We can assume users want to go off the domain in this case
+        baseUrl = finalRequestUrlOrigin ?? originalRequestUrlOrigin;
+    } else if (originalRequestUrlOrigin === finalRequestUrlOrigin) {
+        // If the domains are the same, we can use the original domain
+        baseUrl = originalRequestUrlOrigin;
+    }
```
Comment (B4nan) [inline packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1090]: the code looks almost identical to the one in browser crawler, can we refactor this out? i guess this version would work for the browser crawler too, right?

File: packages/cheerio-crawler/src/internals/cheerio-crawler.ts [hunk 5]
```diff
@@ -1062,23 +1065,39 @@ export class CheerioCrawler<JSONData = unknown> extends BasicCrawler<
...
+    // else {
+    // TODO: Do we want to use finalRequestUrlOrigin if domains match or leave it as undefined?
```
Comment (B4nan) [inline packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1092]: not sure about this, can you give me an example on some edge cases where this matters?

**Review by B4nan — CHANGES_REQUESTED** [review]

File: packages/cheerio-crawler/src/internals/cheerio-crawler.ts (hunk 5 above)
Comment (vladfrangu) [reply packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1092]: The only time I could think this would be an edge case is `domain.com/redirect` -> `subdomain.domain.com` but I'm not too sure... That would not have a baseUrl set so no links would be enqueued

File: test/e2e/test-cheerio-enqueue-links/test.mjs (hunk 2 above)
Comment (B4nan) [reply test/e2e/test-cheerio-enqueue-links/test.mjs:10]: yeah maybe do it separately, we can merge that right ahead

File: packages/basic-crawler/src/internals/basic-crawler.ts (hunk 1 above)
Comment (vladfrangu) [reply packages/basic-crawler/src/internals/basic-crawler.ts:912]: For basic crawlers only? I agree, but for the rest I think they should stay separate (+ makes testing easier)

File: packages/cheerio-crawler/src/internals/cheerio-crawler.ts (hunk 5 above)
Comment (B4nan) [reply packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1092]: ok, will leave it up to you, i guess we dont need to handle all edge cases right now, especially when we are not sure they exist. we will catch issues like this when we start testing on platform and in real projects.

File: packages/basic-crawler/src/internals/basic-crawler.ts (hunk 1 above)
Comment (B4nan) [reply packages/basic-crawler/src/internals/basic-crawler.ts:912]: i would inline the usage of `basicCrawlerEnqueueLinks`, rest can stay the same (although i would like to not export those directly, but we can deal with that later, there are more candidates that could live in some `internals.js` barrel that would be listed in package exports, not won't be documented)

File: packages/cheerio-crawler/src/internals/cheerio-crawler.ts (hunk 5 above)
Comment (vladfrangu) [reply packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1092]: Should we leave the TODO for now so we don't forget or clean it up and handle it when issues actually arise?

Comment (B4nan) [reply packages/cheerio-crawler/src/internals/cheerio-crawler.ts:1092]: i would either keep it uncommented or remove it fully (together with the unused dependency, unless its used elsewhere too) 

maybe better to have the baseUrl inferred all the time, i am sure we will catch it early if it would be problematic, so i'd be in favour of uncommenting

File: packages/core/src/enqueue_links/enqueue_links.ts [hunk 6]
```diff
@@ -262,3 +262,45 @@ export async function enqueueLinks(options: EnqueueLinksOptions): Promise<BatchA
 
     return requestQueue.addRequests(requests);
 }
+
+export function resolveBaseUrl({
```
Comment (B4nan) [inline packages/core/src/enqueue_links/enqueue_links.ts:266]: let's mark this as internal, otherwise ready to merge

**Review by szmarczak — APPROVED** [review]

**Review by B4nan — APPROVED** [review]
